src/rust/bitbox02-rust/src/hww/api/system.rs

```diff
@@ -0,0 +1,86 @@
+// Copyright 2021 Shift Crypto AG
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use super::Error;
+use crate::pb;
+
+use pb::reboot_request::Purpose;
+use pb::response::Response;
+
+use crate::workflow::confirm;
+
+pub async fn reboot(&pb::RebootRequest { purpose }: &pb::RebootRequest) -> Result<Response, Error> {
+    confirm::confirm(&confirm::Params {
+        title: "",
+        body: match Purpose::from_i32(purpose) {
+            Some(Purpose::Upgrade) => "Proceed to upgrade?",
+            Some(Purpose::Settings) => "Go to\nstartup settings?",
+            // No error, if new client library sends a purpose that we don't understand,
+            // we reboot anyway.
+            None => "Reboot?",
+        },
+        ..Default::default()
+    })
+    .await?;
+    bitbox02::reboot();
+    Ok(Response::Success(pb::Success {}))
```

Comment:
So, if the `bitbox02::reboot` was marked as no return, this line wouldn't be needed. It's unreachable anyway, right.

I'm still unclear on what's this mysterious delay are we talking about? in bitbox02::reboot() or somewhere else? and why would it impact the no return?

Reply:
I just said if hypothetically there was a delay, cause I don't know what is going on in a reboot and how it is performed exactly. If you are sure there is no delay at all, I could mark it no_return.

Still fail to see the benefit to be honest. I'll try doing it anyway, but all this (also marking the FFI wrapper as no return) just to remove this one line...

Reply:
Personally to me it's not "just to remove this one line". It's also about readability. If I see `!` in the funtion declaration I trust it won't return. But at the moment, I look at this:

```rust
bitbox02::reboot();
Ok(Response::Success(pb::Success {}))
```

and my first thought is: wait a second, does reboot returns? when does it actually reboot, in a few microseconds, a minute... an hour? and what happens meanwhile? 

I think I'm starting to understand now. If by "delay" you mean that the `_reboot` > `__NVIC_SystemReset` call chain can return before the MCU is restarted, I just don't see how. https://github.com/digitalbitbox/bitbox02-firmware/pull/739#issuecomment-815617645

Reply:
If you really don't trust the code in C, add an infinite noop loop in `bitbox02::reboot` func in rust?

Reply:
Done. I actually had to add the loop{}, as Rust FFI does not recognize that the wrapped function never returns.